File: web-next/src/libs/api/blogpost.ts
import { ResponseEnum } from "@/libs/enum";
import { API_URL } from "@/libs/config";
import { BlogPost } from "@/libs/types";

type ReturnCountData = {
  data: number;
  success: boolean;
};

type ReturnData = {
  data: any;
  success: boolean;
};

async function Count(search: string): Promise<ReturnCountData> {
  try {
    const response = await fetch(
      `${API_URL}/blog/posts/count?search=${search}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
        credentials: "include",
      },
    );

    if (!response.ok) {
      return {
        data: 0,
        success: false,
      };
    }

    const jsonResponse = await response.json();

    if (jsonResponse.message === ResponseEnum.GET_DATA_SUCCESS) {
      return {
        data: jsonResponse.data,
        success: true,
      };
    } else {
      return {
        data: 0,
        success: false,
      };
    }
  } catch (error) {
    return {
      data: 0,
      success: false,
    };
  }
}

async function GetAll(
  search: string,
  tags: string,
  limit: number,
  page: number,
): Promise<ReturnData> {
  try {
    const response = await fetch(
      `${API_URL}/blog/posts?search=${search}&tags=${tags}&limit=${limit}&page=${page}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
        credentials: "include",
      },
    );

    if (!response.ok) {
      return {
        data: [],
        success: false,
      };
    }

    const jsonResponse = await response.json();

    if (jsonResponse.message === ResponseEnum.GET_DATA_SUCCESS) {
      return {
        data: jsonResponse.data,
        success: true,
      };
    } else {
      return {
        data: [],
        success: false,
      };
    }
  } catch (error) {
    return {
      data: [],
      success: false,
    };
  }
}

async function Create(input: BlogPost): Promise<ReturnData> {
  try {
    const response = await fetch(`${API_URL}/blog/posts`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      credentials: "include",
      body: JSON.stringify({
        title: input.title,
        slug: input.slug,
        created_at: input.created_at,
        updated_at: input.updated_at,
        is_draft: input.is_draft
      }),
    });

    if (!response.ok) {
      return {
        data: {},
        success: false,
      };
    }

    const jsonResponse = await response.json();

    if (jsonResponse.message === ResponseEnum.CREATE_DATA_SUCCESS) {
      return {
        data: jsonResponse.data,
        success: true,
      };
    } else {
      return {
        data: {},
        success: false,
      };
    }
  } catch (error) {
    return {
      data: {},
      success: false,
    };
  }
}

async function Update(input: BlogPost): Promise<ReturnData> {
  try {
    const response = await fetch(`${API_URL}/blog/posts`, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
      credentials: "include",
      body: JSON.stringify({
        id: input.id,
        title: input.title,
        slug: input.slug,
        created_at: input.created_at,
        updated_at: input.updated_at,
        is_draft: input.is_draft
      }),
    });

    if (!response.ok) {
      return {
        data: {},
        success: false,
      };
    }

    const jsonResponse = await response.json();

    if (jsonResponse.message === ResponseEnum.UPDATE_DATA_SUCCESS) {
      return {
        data: jsonResponse.data,
        success: true,
      };
    } else {
      return {
        data: {},
        success: false,
      };
    }
  } catch (error) {
    return {
      data: {},
      success: false,
    };
  }
}

async function Remove(id: string): Promise<ReturnData> {
  try {
    const response = await fetch(`${API_URL}/blog/posts/${id}`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
      },
      credentials: "include",
    });

    if (!response.ok) {
      return {
        data: {},
        success: false,
      };
    }

    const jsonResponse = await response.json();

    if (jsonResponse.message === ResponseEnum.REMOVE_DATA_SUCCESS) {
      return {
        data: jsonResponse.data,
        success: true,
      };
    } else {
      return {
        data: {},
        success: false,
      };
    }
  } catch (error) {
    return {
      data: {},
      success: false,
    };
  }
}

export default {
  Count,
  GetAll,
  Create,
  Update,
  Remove,
};

File: api-chi/cmd/controllers/blogpost.go
package controllers


import (
	"api-chi/cmd/models"
	"api-chi/cmd/services"
	"api-chi/internal/message"

	"encoding/json"
	"log"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"github.com/go-chi/render"
)

type BlogPostController struct {
	service services.BlogPostService
}

func (c *BlogPostController) Count(w http.ResponseWriter, r *http.Request) {
	// Retrieve query parameters
	search := r.URL.Query().Get("search")

	// Open and close database after end
	err := c.service.Open()
	defer c.service.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Execute Count and return if failed or success
	data, err := c.service.Count(search)
	if err != nil {
		render.Status(r, http.StatusInternalServerError)
		render.JSON(w, r, message.Response{
			Message: message.GET_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	render.Status(r, http.StatusOK)
	render.JSON(w, r, message.Response{
		Message: message.GET_DATA_SUCCESS,
		Data:    data,
	})
}

func (c *BlogPostController) GetAll(w http.ResponseWriter, r *http.Request) {
	// Retrieve query parameters
	search := r.URL.Query().Get("search")
    tags := r.URL.Query().Get("tags")
	limit, err := strconv.Atoi(r.URL.Query().Get("limit"))
	if err != nil {
		render.Status(r, http.StatusBadRequest)
		render.JSON(w, r, message.Response{
			Message: message.GET_DATA_FAILED,
			Data:    nil,
		})
		return
	}
	page, err := strconv.Atoi(r.URL.Query().Get("page"))
	if err != nil {
		render.Status(r, http.StatusBadRequest)
		render.JSON(w, r, message.Response{
			Message: message.GET_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	// Open and close database after end
	err = c.service.Open()
	defer c.service.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Execute Count and return if failed or success
	data, err := c.service.GetAll(search, tags, limit, page)
	if err != nil {
		render.Status(r, http.StatusInternalServerError)
		render.JSON(w, r, message.Response{
			Message: message.GET_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	render.Status(r, http.StatusOK)
	render.JSON(w, r, message.Response{
		Message: message.GET_DATA_SUCCESS,
		Data:    data,
	})
}

func (c *BlogPostController) Create(w http.ResponseWriter, r *http.Request) {
	// Get JSON from user input
	input := models.BlogPost{}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		render.Status(r, http.StatusBadRequest)
		render.JSON(w, r, message.Response{
			Message: message.CREATE_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	// Open and close database after end
	err := c.service.Open()
	defer c.service.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Execute Count and return if failed or success
	data, err := c.service.Create(&input)
	if err != nil {
		render.Status(r, http.StatusInternalServerError)
		render.JSON(w, r, message.Response{
			Message: message.CREATE_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	render.Status(r, http.StatusOK)
	render.JSON(w, r, message.Response{
		Message: message.CREATE_DATA_SUCCESS,
		Data:    data,
	})
}

func (c *BlogPostController) Update(w http.ResponseWriter, r *http.Request) {
	// Get JSON from user input
	input := models.BlogPost{}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		render.Status(r, http.StatusBadRequest)
		render.JSON(w, r, message.Response{
			Message: message.UPDATE_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	// Open and close database after end
	err := c.service.Open()
	defer c.service.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Execute Count and return if failed or success
	data, err := c.service.Update(&input)
	if err != nil {
		render.Status(r, http.StatusInternalServerError)
		render.JSON(w, r, message.Response{
			Message: message.UPDATE_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	render.Status(r, http.StatusOK)
	render.JSON(w, r, message.Response{
		Message: message.UPDATE_DATA_SUCCESS,
		Data:    data,
	})
}

func (c *BlogPostController) Remove(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	if id == "" {
		render.Status(r, http.StatusBadRequest)
		render.JSON(w, r, message.Response{
			Message: message.REMOVE_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	// Open and close database after end
	err := c.service.Open()
	defer c.service.Close()
	if err != nil {
		log.Fatal(err)
	}

	// Execute Count and return if failed or success
	data, err := c.service.Remove(id)
	if err != nil {
		render.Status(r, http.StatusInternalServerError)
		render.JSON(w, r, message.Response{
			Message: message.REMOVE_DATA_FAILED,
			Data:    nil,
		})
		return
	}

	render.Status(r, http.StatusOK)
	render.JSON(w, r, message.Response{
		Message: message.REMOVE_DATA_SUCCESS,
		Data:    data,
	})
}
